android-record/src/main/java/com/fasterxml/jackson/module/androidrecord/AndroidRecordModule.java

```diff
@@ -0,0 +1,95 @@
+package com.fasterxml.jackson.module.androidrecord;
+
+import com.fasterxml.jackson.databind.BeanDescription;
+import com.fasterxml.jackson.databind.DeserializationConfig;
+import com.fasterxml.jackson.databind.PropertyName;
+import com.fasterxml.jackson.databind.deser.CreatorProperty;
+import com.fasterxml.jackson.databind.deser.SettableBeanProperty;
+import com.fasterxml.jackson.databind.deser.ValueInstantiator;
+import com.fasterxml.jackson.databind.deser.std.StdValueInstantiator;
+import com.fasterxml.jackson.databind.introspect.AnnotatedConstructor;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+
+/**
+ * Module that allows deserialization into records
+ * using the canonical constructor on Android,
+ * where java records are supported through desugaring,
+ * and Jackson's built-in support for records doesn't work,
+ * since the desugared classes have a non-standard super class,
+ * and record component-related reflection methods are missing.
+ *
+ * <p>
+ * See <a href="https://android-developers.googleblog.com/2023/06/records-in-android-studio-flamingo.html">
+ *   Android Developers Blog article</a>
+ *
+ * <p>
+ * Note: this module is a no-op when no Android-desugared records are being deserialized,
+ * so it is safe to use in code shared between Android and non-Android platforms.
+ *
+ * <p>
+ * Note: The canonical record constructor is found
+ * through matching of parameter types with field types.
+ * Therefore, this module doesn't allow a deserialized desugared record class to have a custom
+ * constructor with a signature that's any permutation of the canonical one's.
+ *
+ * @author Eran Leshem
+ **/
+public class AndroidRecordModule extends SimpleModule {
+  @Override
+  public void setupModule(SetupContext context) {
+    super.setupModule(context);
+    context.addValueInstantiators(AndroidRecordModule::findValueInstantiator);
+  }
+
+  private static ValueInstantiator findValueInstantiator(DeserializationConfig config, BeanDescription beanDesc,
+                                                         ValueInstantiator defaultInstantiator) {
+    Class<?> raw = beanDesc.getType().getRawClass();
+    if (defaultInstantiator instanceof StdValueInstantiator && raw.getSuperclass() != null
+            && raw.getSuperclass().getName().equals("com.android.tools.r8.RecordTag")) {
+      Map<Type, Integer> componentTypes = typeMap(Arrays.stream(raw.getDeclaredFields())
+              .filter(field -> !Modifier.isStatic(field.getModifiers())).map(Field::getGenericType));
+      boolean found = false;
+      for (Constructor<?> constructor: raw.getDeclaredConstructors()) {
```

Comment:
For constructors it'd be better to access `AnnotatedConstructor`s from `beanDesc` if at all accessible.
Same goes for fields (-> `AnnotatedField`).
These would give you instances needed for constructing other objects, as well as processed annotations (including possible mix-ins).

Reply:
> For constructors it'd be better to access `AnnotatedConstructor`s from `beanDesc` if at all accessible. 

Done, thanks.

> Same goes for fields (-> `AnnotatedField`). 

Couldn't find that.

Reply:
Ah. Should have looked at it: method to look for is

    public abstract AnnotatedClass getClassInfo();

and `AnnotatedClass` has method `fields()` to get them.

Reply:
Thanks. Unfortunately, it seems to not be useful here. This module uses three pieces of metadata about fields: name, modifiers, and generic type. AnnotatedField is a pass-through for the former two, and the latter is unavailable on it AFAICT.